### PR TITLE
send full dataslot value on neo<->neo connections

### PR DESCRIPTION
**Pull request by justliliandev**
## The Problem
Dataslots send only the short value over the network, this is enough for all vanilla use cases, however some modders want to synchronize higher values using dataslots. The problem is also annoying to catch as some modders don't test their full functionality on a dedicated servers and only there the information is lost. Some modders made their own helpers to have 2 dataslots and use bitmanipulation to set a single integer. That's needlessly complex.

## Solution
Send the full 32bit integer of a dataslot over the network on neo<->neo connections, so that modders don't have to use multiple dataslots for a piece of data and to prevent/fix synchronization errors in mods

## Testing
I've verified that this PR works on an integrated server and on a dedicated neo server with a neo and vanilla client.

### CODE REVIEW

**Comment by Shadows-of-Fire** [comment]
Tbh this is huge, I can drop so many stupid workarounds now.

**Review by Technici4n — COMMENTED** [review]
Not a bad idea. Unfortunately our higher furnace burn time will still cause artifacts on vanilla clients if we only do this.

**Comment by Shadows-of-Fire** [comment]
It always will - the solution is to not use values outside of vanilla's scope if the server is allowing vanilla clients.  The server controls what data files are loaded, after all.

We could potentially log a warning if the server would be otherwise vanilla compatible.

**Comment by Technici4n** [comment]
@Shadows-of-Fire we can tweak the data slots to make the display correct on a vanilla client.

**Comment by justliliandev** [comment]
what do you mean with tweaking the dataslots? we can't tweak them for anvil cost as the data is displayed directly. we could do it for smelting progress and fuel progress, but I'm not sure if that should be part of this PR

**Comment by Technici4n** [comment]
it can be left for another PR yes (if desired at all, but honestly I don't care too much about making our extensions usable by vanilla clients)

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java [hunk 1]
```diff
@@ -717,7 +718,17 @@ public boolean onVanillaNetworkConnectionEstablished(ClientConfigurationPacketLi
...
+        return isVanillaConnection(connection.channel());
+    }
+
+    /**
+     * Indicates if the given connection is a vanilla connection.
+     *
+     * @param channel The channel to check.
+     * @return True if the connection is a vanilla connection, false otherwise.
+     */
+    public boolean isVanillaConnection(Channel channel) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java]: Why was this added?

File: src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import net.minecraft.network.protocol.game.ClientboundContainerSetDataPacket;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.internal.versions.neoforge.NeoForgeVersion;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * A custom payload that updates the full dataslot value instead of just the short value
+ *
+ * @param containerId The containerId for the container.
+ * @param dataId      The id for the dataslot.
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java]:
```suggestion
 * @param dataId      The ID of the dataslot.
```

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java (hunk 1 above)
Comment (justliliandev) [reply src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java]: This is an artifact of another approach I tried first, it will be removed

File: src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * A custom payload that updates the full dataslot value instead of just the short value
+ *
+ * @param containerId The containerId for the container.
+ * @param dataId      The ID of the dataslot.
+ * @param value       The value of the dataslot.
+ */
+@ApiStatus.Internal
+public record AdvancedContainerSetDataPayload(int containerId, int dataId, int value) implements CustomPacketPayload {
+
+    public static final ResourceLocation ID = new ResourceLocation(NeoForgeVersion.MOD_ID, "advanced_set_data");
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java]:
```suggestion
    public static final ResourceLocation ID = new ResourceLocation(NeoForgeVersion.MOD_ID, "advanced_set_container_data");
```

Comment (justliliandev) [reply src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java]: I'm fine with adding container to the string, but shouldn't it then be `advanced_container_set_data` to match the class name?

Comment (Matyrobbrt) [reply src/main/java/net/neoforged/neoforge/network/payload/AdvancedContainerSetDataPayload.java]: Sure, just add container.

**Review by Matyrobbrt — APPROVED** [review]
